### PR TITLE
Web3

### PR DESCRIPTION
**Pull request by dudiq**

### CODE REVIEW

**Comment by dudiq** [comment]
@kazurus

File: packages/web3/src/Connections/WsConnection.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        console.error(e);
+      }
+    };
+  }
+
+  /**
+   * @param {object} data
+   * @return {Promise<void>}
+   */
+  async sendViaWs(data) {
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: send?

File: packages/web3/src/Connections/WsConnection.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      this.reconnect();
+    };
+    ws.onmessage = ({ data }) => {
+      try {
+        const object = JSON.parse(data);
+        if (object.id) {
+          // :TODO rethink about detect event and request
+          this.handleRequest(object);
+        } else {
+          this.handleEvent(object);
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: handleSubsctiptionEvent?

File: packages/web3/src/Connections/WsConnection.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+   */
+  reconnect() {
+    this.initWebSocket(RECONNECT_TIMEOUT);
+  }
+
+  /**
+   * @private
+   * @param {Function} onCreate
+   */
+  createNativeWS(onCreate) {
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: createConnection?

File: packages/web3/src/Connections/WsConnection.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    const ws = new window.WebSocket(this.url);
+    this.ws = ws;
+    ws.binaryType = 'arraybuffer';
+
+    ws.onopen = () => {
+      this.isReady = true;
+      onCreate();
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: handleCreate

File: packages/web3/src/Connections/WsConnection.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      await this.resolvePromise;
+    }
+
+    if (!this.ws) {
+      throw new Error('WebSocket is not init, something goes wrong');
+    }
+
+    const str = JSON.stringify(data);
+
+    this.ws.send(str);
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: this.connection.send

File: packages/web3/src/Connections/WsConnection.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    this.isDestroyed = false;
+    this.reconnectTimerId = undefined;
+  }
+
+  /**
+   * @private
+   * @param {number} timeout
+   * @return {Promise<void>}
+   */
+  initWebSocket(timeout = 0) {
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: init

File: packages/web3/src/Connections/WsConnection.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    const str = JSON.stringify(data);
+
+    this.ws.send(str);
+  }
+
+  /**
+   * @param {object} data
+   * @return {void}
+   */
+  send = data => {
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: do we need it?

File: packages/web3/src/Connections/WsConnection.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  async create() {
+    if (this.isReady) {
+      return this.resolvePromise;
+    }
+
+    return this.initWebSocket();
+  }
+
+  destroy() {
+    this.isDestroyed = true;
```
Comment (kazurus) [inline packages/web3/src/Connections/WsConnection.js]: do we need destroy connection?

File: packages/web3/src/Events/BaseEvent.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
+export default class BaseEvent {
+  static eventName = 'not_defined_event';
+
+  constructor({ context }) {
+    this.context = context;
+
+    /** @type {EventHashMap} */
+    this.hashMap = {};
```
Comment (kazurus) [inline packages/web3/src/Events/BaseEvent.js]: eventsHashMap?

File: packages/web3/src/Events/BaseEvent.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        delete hashMap[hashKey];
+      }
+    });
+
+    if (Object.keys(hashMap).length === 0) {
+      this.releaseCallbacks();
+    }
+  }
+
+  getHash(data) {
```
Comment (kazurus) [inline packages/web3/src/Events/BaseEvent.js]: get hash

File: packages/web3/src/class/NotifyHandler.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+   */
+  subscribe(cb) {
+    this.callbacks.push(cb);
+  }
+
+  /**
+   * @param {Function=} cb
+   */
+  unsubscribe(cb) {
+    if (!cb) {
```
Comment (kazurus) [inline packages/web3/src/class/NotifyHandler.js]: destroy?

File: packages/web3/src/plugins/Balance/BalanceEvent.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  constructor(props) {
+    super(props);
+    this.requestMap = {};
+    this.isSkipping = false;
+    this.subscribeId = null;
+    this.isSubscribed = false;
+  }
+
+  getHash({ address }) {
```
Comment (kazurus) [inline packages/web3/src/plugins/Balance/BalanceEvent.js]: ?

File: packages/web3/src/plugins/Composite.js [hunk 13]
```diff
@@ -0,0 +1,3 @@
+import BasePlugin from './BasePlugin';
+
+export default class CompositePlugin extends BasePlugin {}
```
Comment (kazurus) [inline packages/web3/src/plugins/Composite.js]: name?

**Pull request merged**